// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Protobuf messages for arolla array serialization.

syntax = "proto2";

package arolla.serialization_codecs;

import "arolla/serialization_base/base.proto";

option cc_enable_arenas = true;

message ArrayV1Proto {
  extend arolla.serialization_base.ValueProto {
    optional ArrayV1Proto extension = 361119001;
  }

  // Represents Array<T>.
  //
  // ValueProto.input_value_indices[0]
  //   -- (optional) Stores a DenseArray<T> dense_data,
  //      iff 0 < size.
  //
  // ValueProto.input_value_indices[1]
  //   -- (optional) Stores an OptionalValue<T> missing_id_value,
  //      iff 0 <= dense_data.size() < size.
  //
  message ArrayProto {
    optional int64 size = 1;
    repeated int64 ids = 2 [packed = true];
  }

  // Represents ArrayEdge.
  //
  // ValueProto.input_value_indices[0]
  //   -- Stores a Array<int64_t> with split_points or mapping.
  message ArrayEdgeProto {
    enum EdgeType {
      EDGE_TYPE_UNSPECIFIED = 0;
      MAPPING = 1;
      SPLIT_POINTS = 2;
    }
    optional EdgeType edge_type = 1;

    // Only used when edge_type = MAPPING (for SPLIT_POINTS,
    // ValueProto.input_value_indices[0] represents all the required
    // information).
    optional int64 parent_size = 2;
  }

  reserved 5, 6, 105, 106;
  oneof value {
    ArrayProto array_unit_value = 1;
    ArrayProto array_boolean_value = 2;
    ArrayProto array_bytes_value = 3;
    ArrayProto array_text_value = 4;
    ArrayProto array_int32_value = 7;
    ArrayProto array_int64_value = 8;
    ArrayProto array_uint64_value = 9;
    ArrayProto array_float32_value = 10;
    ArrayProto array_float64_value = 11;

    ArrayEdgeProto array_edge_value = 50;
    int64 array_to_scalar_edge_value = 51;
    int64 array_shape_value = 52;

    bool array_unit_qtype = 101;
    bool array_boolean_qtype = 102;
    bool array_bytes_qtype = 103;
    bool array_text_qtype = 104;
    bool array_int32_qtype = 107;
    bool array_int64_qtype = 108;
    bool array_uint64_qtype = 109;
    bool array_float32_qtype = 110;
    bool array_float64_qtype = 111;

    bool array_edge_qtype = 150;
    bool array_to_scalar_edge_qtype = 151;
    bool array_shape_qtype = 152;
  }
}
